web/src/test/auto/headless/engine/predictive-text/worker-thread/context/context-state.tests.ts

```diff
@@ -55,6 +69,282 @@ describe('ContextState', () => {
       assert.isOk(state.tokenization);
       assert.equal(state.tokenization.tokens.length, 9);
       assert.deepEqual(state.tokenization.exampleInput, ['the', ' ', 'quick', ' ', 'brown', ' ', 'fox', ' ', '']);
+
+      let context2 = { left: "an apple a day keeps the doctor ", startOfBuffer: true, endOfBuffer: true };
+      let rawTokens = ["an", " ", "apple", " ", "a", " ", "day", " ", "keeps", " ", "the", " ", "doctor", " ", ""];
+
+      let state2 = new ContextState(context2, plainModel);
+      assert.deepEqual(state2.tokenization.tokens.map(token => token.exampleInput), rawTokens);
+    });
+  });
+
+
+  describe('analyzeTransition', function() {
+    it("properly matches and aligns when lead token is removed", function() {
+      let existingContext = {
+        left: "an apple a day keeps the doctor", startOfBuffer: true, endOfBuffer: true
+      };
+      let transform = {
+        insert: '',
+        deleteLeft: 0
+      };
+      let newContext = {
+        left: " apple a day keeps the doctor", startOfBuffer: true, endOfBuffer: true
+      };
+      let rawTokens = [" ", "apple", " ", "a", " ", "day", " ", "keeps", " ", "the", " ", "doctor"];
+
+      let baseState = new ContextState(existingContext, plainModel);
+      let newContextMatch = baseState.analyzeTransition(newContext, toWrapperDistribution(transform));
+      assert.isNotNull(newContextMatch?.final);
+      assert.deepEqual(newContextMatch.final.tokenization.tokens.map(token => token.exampleInput), rawTokens);
+
+      if(!newContextMatch.final.tokenization.alignment.canAlign) {
+        assert.fail("context alignment failed");
+      }
```

Comment:
Or:

```suggestion
      assert.isTrue(newContextMatch.final.tokenization.alignment.canAlign, "context alignment failed");
```

(more occurrences below)

Reply:
> `assert.isTrue` in a check like that doesn't let TS realize that any code following _cannot_ have `.canAlign` false.

This is because `assert.isTrue` is not control flow, so TS would need to have special knowledge that it doesn't return in the case of failure, or deep analysis of control flow, and that's really outside the scope of the compiler.

This also means that @ermshiperete's suggestion wouldn't work either -- it still isn't control flow, so TS cannot infer that value of `canAlign` after the `assert` call.